src/app/application/services/data-source/CallableEventSource.ts

```diff
@@ -0,0 +1,54 @@
+export const isClosed = (source: { readyState: number }) => source.readyState === 2
+export default class CallableEventSource extends EventTarget {
+  url = ''
+  withCredentials = false
+  readonly CONNECTING = 0
+  readonly OPEN = 1
+  readonly CLOSED = 2
+
+  onerror = null
+  onmessage = null
+  onopen = null
+  readyState = 2
+
+  constructor(
+    protected source: () => AsyncGenerator,
+    _configuration = {},
+  ) {
+    super()
+    this._open()
+  }
+
+  _open() {
+    (async function (self) {
+      try {
+        self.readyState = 0
+        const source = self.source()
+        for await (const res of source) {
+          self.dispatchEvent(new MessageEvent('message', {
+            data: res,
+          }))
+          if (self.readyState === 2) {
```

Comment:
**Suggestion**: Please use constants in the places you check against `readyState`. It’s too easy to forget what’s "0", "1", or "2".